storage-proofs/src/vde.rs

```diff
@@ -85,31 +85,46 @@ where
     let key = create_key::<H>(replica_id, v, &parents, &data, graph.degree())?;
     let node_data = H::Domain::try_from_bytes(&data_at_node(data, v)?)?;
 
-    // TODO: round constant
     Ok(H::sloth_decode(&key, &node_data, sloth_iter))
 }
 
-pub fn decode_domain_block<'a, H>(
-    degree: usize,
+pub fn decode_domain_block<H>(
     sloth_iter: usize,
-    replica_id: &'a H::Domain,
-    data: &'a [H::Domain],
-    v: usize,
-    parents: Vec<usize>,
+    replica_id: &H::Domain,
+    data: &[H::Domain],
+    node: usize,
+    parents: &[usize],
 ) -> Result<H::Domain>
 where
     H: Hasher,
 {
-    let byte_data = data
-        .iter()
-        .flat_map(H::Domain::into_bytes)
-        .collect::<Vec<u8>>();
+    let key = create_key_domain::<H>(replica_id, node, parents, data)?;
 
-    let key = create_key::<H>(replica_id, v, &parents, &byte_data, degree)?;
-    let node_data = data[v];
+    Ok(H::sloth_decode(&key, &data[node], sloth_iter))
+}
 
-    // TODO: round constant
-    Ok(H::sloth_decode(&key, &node_data, sloth_iter))
+/// Creates the encoding key, using domain encoded data.
+fn create_key_domain<H: Hasher>(
+    id: &H::Domain,
+    node: usize,
+    parents: &[usize],
+    data: &[H::Domain],
+) -> Result<H::Domain> {
+    let mut hasher = Blake2s::new().hash_length(32).to_state();
+    hasher.update(id.as_ref());
+
+    for parent in parents.iter() {
+        // special super shitty case
+        // TODO: unsuck
+        if node == parents[0] {
+            // skip, as we would only write 0s
+        } else {
```

Comment:
Can't we just leave this here? This edge case doesn't matter for performance at scale, and it's a wart we'll have to keep in sync as things change. Unless 'm missing something, I think we should remove.

Reply:
I think I tried to remove it once and things started to break, lets do this another day